File: backend/tools/nest-single-aggregate-enums.py
#!/usr/bin/env python3
"""Nest single-aggregate *Status enums into their owning aggregate as inner enum Status."""

from __future__ import annotations

import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1] / "src"

NEST = {
    "OnboardingStatus": ("OnboardingApplication", "operator/domain/OnboardingApplication.java"),
    "ReferralStatus": ("ReferralBinding", "settlement/domain/ReferralBinding.java"),
    "BatchStatus": ("SettlementBatch", "settlement/domain/SettlementBatch.java"),
    "AccrualStatus": ("ProfitShareAccrual", "settlement/domain/ProfitShareAccrual.java"),
    "UserCouponStatus": ("UserCoupon", "mall/domain/UserCoupon.java"),
    "MerchantStatus": ("MerchantProfile", "mall/domain/MerchantProfile.java"),
    "MallSkuStatus": ("MallSku", "mall/domain/MallSku.java"),
    "MallOrderStatus": ("MallOrder", "mall/domain/MallOrder.java"),
    "CampaignStatus": ("Campaign", "mall/domain/Campaign.java"),
    "StatementStatus": ("BillingStatement", "credit/domain/BillingStatement.java"),
    "DebtStatus": ("CreditLedgerDebt", "credit/domain/CreditLedgerDebt.java"),
    "CreditStatus": ("CreditProfile", "credit/domain/CreditProfile.java"),
    "OverrideStatus": ("PackageOverride", "operator/domain/PackageOverride.java"),
    "TemplateStatus": ("PackageTemplate", "operator/domain/PackageTemplate.java"),
    "EntitlementStatus": ("Entitlement", "commerce/domain/Entitlement.java"),
    "UsageEventStatus": ("UsageEvent", "commerce/domain/UsageEvent.java"),
    "BatteryAssetStatus": ("BatteryAsset", "commerce/domain/BatteryAsset.java"),
    "OrderStatus": ("Order", "commerce/domain/Order.java"),
    "ProductStatus": ("Product", "commerce/domain/Product.java"),
    "OrgStatus": ("Organization", "operator/domain/Organization.java"),
}


def java_files() -> list[Path]:
    files: list[Path] = []
    for base in (ROOT / "main", ROOT / "test"):
        if base.exists():
            files.extend(base.rglob("*.java"))
    return files


def parse_enum(path: Path, enum_name: str) -> tuple[str, str]:
    text = path.read_text(encoding="utf-8")
    doc = ""
    doc_m = re.search(r"/\*\*(.*?)\*/\s*public enum", text, re.DOTALL)
    if doc_m:
        inner = doc_m.group(1).strip()
        lines = [(" " + ln).rstrip() for ln in inner.splitlines()]
        doc = "    /**\n" + "\n".join(f"     *{ln}" for ln in lines) + "\n     */\n"
    body_m = re.search(r"public enum " + re.escape(enum_name) + r"\s*\{([^}]*)\}", text, re.DOTALL)
    if not body_m:
        raise RuntimeError(f"Cannot parse enum body: {path}")
    return doc, body_m.group(1).strip()


def nest_into_aggregate(aggregate_rel: str, doc: str, body: str) -> None:
    path = Path(str(ROOT / "main" / "java" / "com" / "evolutionary") / Path(aggregate_rel))
    text = path.read_text(encoding="utf-8")
    if re.search(r"public enum Status\s*\{", text):
        return
    simple = path.stem
    marker = f"public final class {simple} {{"
    if marker not in text:
        raise RuntimeError(f"Marker not found in {path}")
    inner_body = "\n".join("        " + ln.rstrip() for ln in body.splitlines() if ln.strip())
    inner = f"{marker}\n\n{doc}    public enum Status {{\n{inner_body}\n    }}\n"
    path.write_text(text.replace(marker, inner, 1), encoding="utf-8")


def replace_refs(old: str, aggregate: str) -> None:
    new = f"{aggregate}.Status"
    import_re = re.compile(
        rf"^import com\.evolutionary\.[\w.]+\.{re.escape(old)};\s*\n", re.MULTILINE
    )
    for path in java_files():
        text = path.read_text(encoding="utf-8")
        updated = import_re.sub("", text)
        updated = updated.replace(old, new)
        if updated != text:
            path.write_text(updated, encoding="utf-8")


def find_enum_file(enum_name: str) -> Path:
    for path in java_files():
        if path.name == f"{enum_name}.java" and "domain" in path.as_posix():
            return path
    raise FileNotFoundError(enum_name)


def main() -> None:
    for old, (aggregate, agg_rel) in NEST.items():
        enum_path = find_enum_file(old)
        doc, body = parse_enum(enum_path, old)
        nest_into_aggregate(agg_rel, doc, body)
        replace_refs(old, aggregate)
        enum_path.unlink()
        print(f"nested {old} -> {aggregate}.Status")


if __name__ == "__main__":
    main()

File: backend/tools/nest-single-aggregate-enums.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, "..", "src");

const NEST = {
  OnboardingStatus: ["OnboardingApplication", "operator/domain/OnboardingApplication.java"],
  ReferralStatus: ["ReferralBinding", "settlement/domain/ReferralBinding.java"],
  BatchStatus: ["SettlementBatch", "settlement/domain/SettlementBatch.java"],
  AccrualStatus: ["ProfitShareAccrual", "settlement/domain/ProfitShareAccrual.java"],
  UserCouponStatus: ["UserCoupon", "mall/domain/UserCoupon.java"],
  MerchantStatus: ["MerchantProfile", "mall/domain/MerchantProfile.java"],
  MallSkuStatus: ["MallSku", "mall/domain/MallSku.java"],
  MallOrderStatus: ["MallOrder", "mall/domain/MallOrder.java"],
  CampaignStatus: ["Campaign", "mall/domain/Campaign.java"],
  StatementStatus: ["BillingStatement", "credit/domain/BillingStatement.java"],
  DebtStatus: ["CreditLedgerDebt", "credit/domain/CreditLedgerDebt.java"],
  CreditStatus: ["CreditProfile", "credit/domain/CreditProfile.java"],
  OverrideStatus: ["PackageOverride", "operator/domain/PackageOverride.java"],
  TemplateStatus: ["PackageTemplate", "operator/domain/PackageTemplate.java"],
  EntitlementStatus: ["Entitlement", "commerce/domain/Entitlement.java"],
  UsageEventStatus: ["UsageEvent", "commerce/domain/UsageEvent.java"],
  BatteryAssetStatus: ["BatteryAsset", "commerce/domain/BatteryAsset.java"],
  OrderStatus: ["Order", "commerce/domain/Order.java"],
  ProductStatus: ["Product", "commerce/domain/Product.java"],
  OrgStatus: ["Organization", "operator/domain/Organization.java"],
};

function walk(dir, out = []) {
  if (!fs.existsSync(dir)) return out;
  for (const name of fs.readdirSync(dir)) {
    const p = path.join(dir, name);
    if (fs.statSync(p).isDirectory()) walk(p, out);
    else if (name.endsWith(".java")) out.push(p);
  }
  return out;
}

function javaFiles() {
  return [...walk(path.join(ROOT, "main")), ...walk(path.join(ROOT, "test"))];
}

function parseEnum(file, enumName) {
  const text = fs.readFileSync(file, "utf8");
  const docM = text.match(/\/\*\*([\s\S]*?)\*\/\s*public enum/);
  let doc = "";
  if (docM) {
    const lines = docM[1].trim().split("\n").map((l) => l.trim());
    doc = "    /**\n" + lines.map((l) => "     * " + l).join("\n") + "\n     */\n";
  }
  const bodyM = text.match(new RegExp(`public enum ${enumName}\\s*\\{([^}]*)\\}`, "s"));
  if (!bodyM) throw new Error(`parse fail ${file}`);
  return { doc, body: bodyM[1].trim() };
}

function nest(aggRel, doc, body) {
  const file = path.join(ROOT, "main", "java", "com", "evolutionary", ...aggRel.split("/"));
  let text = fs.readFileSync(file, "utf8");
  if (/public enum Status\s*\{/.test(text)) return;
  const simple = path.basename(file, ".java");
  const marker = `public final class ${simple} {`;
  if (!text.includes(marker)) throw new Error(`no marker ${file}`);
  const innerBody = body
    .split("\n")
    .filter((l) => l.trim())
    .map((l) => "        " + l.trim())
    .join("\n");
  const inner = `${marker}\n\n${doc}    public enum Status {\n${innerBody}\n    }\n`;
  fs.writeFileSync(file, text.replace(marker, inner));
}

function replaceRefs(old, aggregate) {
  const neu = `${aggregate}.Status`;
  const importRe = new RegExp(`^import com\\.evolutionary\\.[\\w.]+\\.${old};\\s*\\n`, "gm");
  for (const file of javaFiles()) {
    let text = fs.readFileSync(file, "utf8");
    const updated = text.replace(importRe, "").replaceAll(old, neu);
    if (updated !== text) fs.writeFileSync(file, updated);
  }
}

function findEnum(enumName) {
  return javaFiles().find(
    (f) => path.basename(f) === `${enumName}.java` && f.includes(`${path.sep}domain${path.sep}`)
  );
}

for (const [old, [aggregate, aggRel]] of Object.entries(NEST)) {
  const enumFile = findEnum(old);
  if (!enumFile) throw new Error(`missing ${old}`);
  const { doc, body } = parseEnum(enumFile, old);
  nest(aggRel, doc, body);
  replaceRefs(old, aggregate);
  fs.unlinkSync(enumFile);
  console.log(`nested ${old} -> ${aggregate}.Status`);
}
